File: go/test/detection_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/anti-phishing-detection-sdk/go"
	"github.com/voxgig-sdk/anti-phishing-detection-sdk/go/core"

	vs "github.com/voxgig-sdk/anti-phishing-detection-sdk/go/utility/struct"
)

func TestDetectionEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Detection(nil)
		if ent == nil {
			t.Fatal("expected non-nil DetectionEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := detectionBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "detection." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set ANTIPHISHINGDETECTION_TEST_DETECTION_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		detectionRef01Ent := client.Detection(nil)
		detectionRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "detection"}, setup.data), "detection_ref01"))

		detectionRef01DataResult, err := detectionRef01Ent.Create(detectionRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		detectionRef01Data = core.ToMapAny(detectionRef01DataResult)
		if detectionRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LIST
		detectionRef01Match := map[string]any{}

		detectionRef01ListResult, err := detectionRef01Ent.List(detectionRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		detectionRef01List, detectionRef01ListOk := detectionRef01ListResult.([]any)
		if !detectionRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", detectionRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(detectionRef01List), map[string]any{"id": detectionRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

	})
}

func detectionBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "detection", "DetectionTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read detection test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse detection test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"detection01", "detection02", "detection03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("ANTIPHISHINGDETECTION_TEST_DETECTION_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"ANTIPHISHINGDETECTION_TEST_DETECTION_ENTID": idmap,
		"ANTIPHISHINGDETECTION_TEST_LIVE":      "FALSE",
		"ANTIPHISHINGDETECTION_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["ANTIPHISHINGDETECTION_TEST_DETECTION_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["ANTIPHISHINGDETECTION_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewAntiPhishingDetectionSDK(core.ToMapAny(mergedOpts))
	}

	live := env["ANTIPHISHINGDETECTION_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["ANTIPHISHINGDETECTION_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/DetectionEntityTest.php
<?php
declare(strict_types=1);

// Detection entity test

require_once __DIR__ . '/../antiphishingdetection_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class DetectionEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = AntiPhishingDetectionSDK::test(null, null);
        $ent = $testsdk->Detection(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = detection_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "detection." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set ANTIPHISHINGDETECTION_TEST_DETECTION_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $detection_ref01_ent = $client->Detection(null);
        $detection_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.detection"), "detection_ref01"));

        $detection_ref01_data_result = $detection_ref01_ent->create($detection_ref01_data, null);
        $detection_ref01_data = Helpers::to_map($detection_ref01_data_result);
        $this->assertNotNull($detection_ref01_data);

        // LIST
        $detection_ref01_match = [];

        $detection_ref01_list_result = $detection_ref01_ent->list($detection_ref01_match, null);
        $this->assertIsArray($detection_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($detection_ref01_list_result),
            ["id" => $detection_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

    }
}

function detection_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/detection/DetectionTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = AntiPhishingDetectionSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["detection01", "detection02", "detection03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("ANTIPHISHINGDETECTION_TEST_DETECTION_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "ANTIPHISHINGDETECTION_TEST_DETECTION_ENTID" => $idmap,
        "ANTIPHISHINGDETECTION_TEST_LIVE" => "FALSE",
        "ANTIPHISHINGDETECTION_TEST_EXPLAIN" => "FALSE",
        "ANTIPHISHINGDETECTION_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["ANTIPHISHINGDETECTION_TEST_DETECTION_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["ANTIPHISHINGDETECTION_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["ANTIPHISHINGDETECTION_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new AntiPhishingDetectionSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["ANTIPHISHINGDETECTION_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["ANTIPHISHINGDETECTION_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
